### PR TITLE
RatakondalaArun/web-support

### PR DESCRIPTION
**Pull request by RatakondalaArun**
**This PR adds** 
- initial support for web as described in https://github.com/fluttercommunity/flutter_launcher_icons/issues/348#issuecomment-1165773286
- a base system for other platforms
- initial type-safe support for `flutter_launcher_icons.yaml` as discussed in #373
- a Logger to better manage standard, error, verbose, and progress logs e597498
- `--prefix` flag to support running commands in different directories. Introduced for tests b2b8982

**Tests**
- fli config yaml d3c061d
- utils tests c2aaaa8
- generator tests 3e2578e, 96a54a5

**post review todo:**
- [ ] update docs
- [x] link issues that close with this PR

### CODE REVIEW

**Comment by RatakondalaArun** [comment]
cc @MarkOSullivan94, @bradintheusa 
This PR is ready for review 🎉

**Review by bradintheusa — APPROVED** [review]
This looks good. I like what you have done.

**Comment by RatakondalaArun** [comment]
Hi @bradintheusa 

please approve this PR once again.

**Comment by RatakondalaArun** [comment]
Hi @MarkOSullivan94 Please review this PR

**Comment by MarkOSullivan94** [comment]
Reviewing now

File: lib/abs/icon_generator.dart [hunk 1]
```diff
@@ -0,0 +1,90 @@
+import 'dart:io';
+
+import 'package:flutter_launcher_icons/flutter_launcher_icons_config.dart';
+import 'package:flutter_launcher_icons/logger.dart';
+
+///A base class to generate icons
```
Comment (MarkOSullivan94) [inline lib/abs/icon_generator.dart:6]:
```suggestion
/// A base class to generate icons
```

Being picky with this one but space helps make it more readable

File: analysis_options.yaml [hunk 2]
```diff
@@ -13,12 +13,12 @@ analyzer:
     # Please see https://github.com/flutter/flutter/pull/24528 for details.
```
Comment (MarkOSullivan94) [inline analysis_options.yaml:13]: Would like to propose adding three new options

```
    public_member_api_docs: true
    use_super_parameters: true
    require_trailing_commas: true
```

**Why?**

1. `public_member_api_docs` 

Will help encourage contributions to document any changes they make in PRs, found this to be typically really useful in projects I've worked on where I'm looking back at code I wrote 6+ months ago and it's far easier to understand with good, meaningful comments

I wouldn't worry so much to document **everything** in this PR but anything new should be commented and we can gradually document more later

2. `use_super_parameters` 

Always think it's good to adopt the latest features introduced in newer versions of Dart, been using this super parameters in other projects and it helps (imo) make the code a lot cleaner 

3. `require_trailing_commas` 

Another one to help improve readability - it helps avoid code from going on to another line and not being nicely formatted

File: lib/abs/icon_generator.dart [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+  IconGenerator(this.context, this.platformName);
+
+  /// Creates icons for this platform.
+  void createIcons();
+
+  /// Should return `true` if this platform
+  /// has all the requirments to create icons.
+  /// This runs before to [createIcons]
+  bool validateRequirments();
```
Comment (MarkOSullivan94) [inline lib/abs/icon_generator.dart:19]: Typo

```suggestion
  bool validateRequirements();
```

File: lib/abs/icon_generator.dart [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        progress.cancel();
+        logger
+          ..error(e.toString())
+          ..verbose(st);
+        continue;
+      }
+    }
+  } catch (e, st) {
+    // todo: better error handling
+    // stacktrace should only print when verbose is turned on
```
Comment (MarkOSullivan94) [inline lib/abs/icon_generator.dart:83]: All for this 👍 

Feel free to create a ticket for the work required to add this

File: analysis_options.yaml (hunk 2 above)
Comment (RatakondalaArun) [reply analysis_options.yaml:13]: adopting the `use_super_parameters` lint rule may require a dart SDK version bump

File: lib/abs/icon_generator.dart [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  final FlutterLauncherIconsConfig config;
+  final FLILogger logger;
+  final String? flavor;
+  final String prefixPath;
+
+  IconGeneratorContext({
+    required this.config,
+    this.flavor,
+    required this.prefixPath,
+    required this.logger,
```
[the remark this answers is not in the document]
Comment (RatakondalaArun) [reply lib/abs/icon_generator.dart]: fixed in 3a5822e

File: lib/abs/icon_generator.dart (hunk 4 above)
Comment (RatakondalaArun) [reply lib/abs/icon_generator.dart:102]: raised a new ticket for this
- https://github.com/fluttercommunity/flutter_launcher_icons/issues/378

File: lib/web/web_icon_generator.dart [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    _generateIcons(imgFile);
+
+    // update manifest.json in web/mainfest.json
+    context.logger.verbose('Updating ${path.join(context.prefixPath, constants.webManifestFilePath)}...');
+    _updateManifestFile();
+
+    // todo: update index.html in web/index.html
+    // as we are using flutter default config we no need
+    // to update index.html for now
+    // _updateIndexFile();
```
Comment (MarkOSullivan94) [inline lib/web/web_icon_generator.dart:75]: Does this have a ticket?

**Review by MarkOSullivan94 — APPROVED** [review]
LGTM! 👍

**Pull request merged**